uniconnect-epfl/uniconnect
# feat(contact): new contact structure and on-device storage

# What I did
- I implemented the contact type and refactored all code requiring this structure.
- I implemented a simple, asynchronous, unencrypted, persistent, key-value storage system for storing the graph object to avoid recreating one each time. Note that the graph object is deleted at the launch of the app and re-created the first time we navigate to the __Explore__ section.

# How I did it

- I created a Contact type in Contact.tsx and refactored files involved in the __Explore__ section which were all using their own dummy data.
- I used the [AsyncStorage API ](https://react-native-async-storage.github.io/async-storage/)
- I extracted the methods from the Graph class as AsyncStorage can only store __strings__ and functions cannot be **serialized** making them impossible to use when retrieving the Graph object from the storage system.

# How to verify it

To check the AsyncStorage:
1. Navigate to the __Explore__ section and click on the __Graph View__
2. Check that the graph layout does not change when navigating to another part of the app and coming back to the __Graph View__ (given that you did not modify the drag by dragging it or any node)
3. Reload the application, navigate to the __Explore__ section and click on the __Graph View__
4. Check that the graph layout is different from last time (Note that it can still be the same as the node placement algorithm can converge to the same positions as last time - in this case retry this step one or two times)

To check the Contact implementation:
1. Check that contacts, either in the __Plain View__ (i.e., list) or in the __Graph View__ have their own profile picture and that they are displayed correctly
2. In the __Graph View__ check that the further a node is from the **main node** (i.e., the user itself), the smallest it is.

# Demo video
https://github.com/uniconnect-epfl/uniconnect/assets/81919237/4c4a5933-bbe7-499c-974a-29788b95748b

# Pre-merge checklist
The changes I have introduced:
